### ISSUE TITLE
kube-proxy's init container stuck in `Waiting` phase while kube-proxy container is running

### ISSUE THREAD

**Issue by MrHohn** [kubernetes/kubernetes#52462]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

Saw kube-proxy pods stuck in a weird state on a long running cluster as below:
```
NAME             READY     STATUS     RESTARTS   AGE
kube-proxy-xxx   0/1       Init:0/1   0          27d
kube-proxy-xxx   0/1       Init:0/1   0          36d
kube-proxy-xxx   0/1       Init:0/1   0          27d
kube-proxy-xxx   0/1       Init:0/1   0          36d
kube-proxy-xxx   0/1       Init:0/1   0          27d
kube-proxy-xxx   0/1       Init:0/1   0          27d
kube-proxy-xxx   0/1       Init:0/1   0          27d
kube-proxy-xxx   0/1       Init:0/1   0          27d
kube-proxy-xxx   0/1       Init:0/1   0          27d
kube-proxy-xxx   0/1       Init:0/1   0          36d
kube-proxy-xxx   0/1       Init:0/1   0          27d
kube-proxy-xxx   1/1       Running    2          36d
...
```

Logged on those unready kube-proxy pods and found kube-proxy container is in fact running (iptables entries is being updated).

`kubectl describe` the unready kube-proxy pods show init container stuck in `Waiting` phase.
```
Name:           kube-proxy-...
Namespace:      kube-system
Node:           ...
Start Time:     ...
Labels:         component=kube-proxy
                tier=node
Annotations:    ...
                scheduler.alpha.kubernetes.io/critical-pod=
Status:         Pending
IP:             ...
Init Containers:
  touch-lock:
    Container ID:       
    Image:              busybox
    Image ID:           
    Port:               <none>
    Command:
      /bin/touch
      /run/xtables.lock
    State:              Waiting
      Reason:           PodInitializing
    Ready:              False
    Restart Count:      0
    Environment:        <none>
    Mounts:
      ...
Containers:
  kube-proxy:
    Container ID:       ...
    Image:              gcr.io/google_containers/kube-proxy:v1.7.2
    Image ID:           ...
    Port:               <none>
    Command:
      /bin/sh
      -c
      echo -998 > /proc/$$$/oom_score_adj && kube-proxy --master=... --kubeconfig=/var/lib/kube-proxy/kubeconfig --cluster-cidr=... --resource-container="" --v=2 --feature-gates=ExperimentalCriticalPodAnnotation=true --iptables-sync-period=1m --iptables-min-sync-period=10s 1>>/var/log/kube-proxy.log 2>&1
    State:              Running
      Started:          ...
    Ready:              True
    Restart Count:      2
    Requests:
      cpu:              100m
    Environment:        <none>
    Mounts:
      ...
Conditions:
  Type          Status
  Initialized   False 
  Ready         True 
  PodScheduled  True
Volumes:
  ...
QoS Class:      Burstable
Node-Selectors: <none>
Tolerations:    :NoExecute
Events:         <none>
```

**What you expected to happen**:

Init container should finish running and kube-proxy should become ready.

**How to reproduce it (as minimally and precisely as possible)**:

Not sure how it ended up with this state yet.

**Anything else we need to know?**:
@kubernetes/sig-node-bugs 
cc @freehan @dnardo 

**Environment**:
- Kubernetes version (use `kubectl version`): Master 1.7.3, nodes 1.7.2
- Cloud provider or hardware configuration**: GKE, though don't think it's provider specific
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by yujuhong** [comment]
> Logged on those unready kube-proxy pods and found kube-proxy container is in fact running (iptables entries is being updated).

Were you able to login to the node and check all docker containers (`docker ps -a`)? One possible cause is that the exited init containers got removed manually.

**Comment by Jacobh2** [comment]
I'm having the same problem with one of three kube-proxy pods:
```
Name:		kube-proxy-gke-...
Namespace:	kube-system
Node:		gke-...
Start Time:	Thu, 14 Sep 2017 12:38:21 +0200
Labels:		component=kube-proxy
		tier=node
Annotations:	kubernetes.io/config.hash=244f7a66cadd842dfe573a29252ef0ea
		kubernetes.io/config.mirror=244f7a66cadd842dfe573a29252ef0ea
		kubernetes.io/config.seen=2017-09-14T10:38:15.940504138Z
		kubernetes.io/config.source=file
		scheduler.alpha.kubernetes.io/critical-pod=
Status:		Pending
IP:		10.132.0.8
Init Containers:
  touch-lock:
    Container ID:
    Image:		busybox
    Image ID:
    Port:		<none>
    Command:
      /bin/touch
      /run/xtables.lock
    State:		Waiting
      Reason:		PodInitializing
    Ready:		False
    Restart Count:	0
    Environment:	<none>
    Mounts:
      /run from run (rw)
Containers:
  kube-proxy:
    Container ID:	...
    Image:		gcr.io/google_containers/kube-proxy:v1.7.5
    Image ID:		...
    Port:		<none>
    Command:
      /bin/sh
      -c
      echo -998 > /proc/$$$/oom_score_adj && kube-proxy --master=https://<ip> --kubeconfig=/var/lib/kube-proxy/kubeconfig --cluster-cidr=10.36.0.0/14 --resource-container="" --v=2 --feature-gates=ExperimentalCriticalPodAnnotation=true --iptables-sync-period=1m --iptables-min-sync-period=10s 1>>/var/log/kube-proxy.log 2>&1
    State:		Running
      Started:		Thu, 14 Sep 2017 12:38:33 +0200
    Ready:		True
    Restart Count:	0
    Requests:
      cpu:		100m
    Environment:	<none>
    Mounts:
      /etc/ssl/certs from etc-ssl-certs (ro)
      /run/xtables.lock from iptableslock (rw)
      /usr/share/ca-certificates from usr-ca-certs (ro)
      /var/lib/kube-proxy/kubeconfig from kubeconfig (rw)
      /var/log from varlog (rw)
Conditions:
  Type		Status
  Initialized 	False
  Ready 	True
  PodScheduled 	True
Volumes:
  usr-ca-certs:
    Type:	HostPath (bare host directory volume)
    Path:	/usr/share/ca-certificates
  etc-ssl-certs:
    Type:	HostPath (bare host directory volume)
    Path:	/etc/ssl/certs
  kubeconfig:
    Type:	HostPath (bare host directory volume)
    Path:	/var/lib/kube-proxy/kubeconfig
  varlog:
    Type:	HostPath (bare host directory volume)
    Path:	/var/log
  run:
    Type:	HostPath (bare host directory volume)
    Path:	/run
  iptableslock:
    Type:	HostPath (bare host directory volume)
    Path:	/run/xtables.lock
QoS Class:	Burstable
Node-Selectors:	<none>
Tolerations:	:NoExecute
Events:		<none>
```

```
> kubectl version
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.5", GitCommit:"17d7182a7ccbb167074be7a87f0a68bd00d58d97", GitTreeState:"clean", BuildDate:"2017-08-31T09:14:02Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.5", GitCommit:"17d7182a7ccbb167074be7a87f0a68bd00d58d97", GitTreeState:"clean", BuildDate:"2017-08-31T08:56:23Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

No direct connections to the nodes has been made so any container hasn't been removed manually.

**Comment by yujuhong** [comment]
@Jacobh2, could you SSH into the node and run `docker ps -a` to see if the exited init container is still visible?

**Comment by Jacobh2** [comment]
@yujuhong I filtered out all the containers that did not match on "kube-proxy" and ended up with these:
```
$ docker ps -a
CONTAINER ID        IMAGE                                     COMMAND                  CREATED             STATUS              PORTS               NAMES
7c76c0e75e7d        1314eb3ac430                              "/bin/sh -c 'echo ..."   11 days ago         Up 11 days                              k8s_kube-proxy_kube-proxy-gke-<CLUSTERNAME>-default-pool-e103106a-jv06_kube-system_244f7a66cadd842dfe573a29252ef0ea_0
f923c7cb38b5        gcr.io/google_containers/pause-amd64:3.0  "/pause"                 11 days ago         Up 11 days                              k8s_POD_kube-proxy-gke-<CLUSTERNAME>-default-pool-e103106a-jv06_kube-system_244f7a66cadd842dfe573a29252ef0ea_0
```

**Comment by resouer** [comment]
It looks like the the initContainer `touch-lock` didn't even start? Have you tried to search `touch-lock` in `kubelet.log` to see what happened to it?

**Comment by Jacobh2** [comment]
I'm not 100% sure which `kubelet.log` file you mean. If it is a file that should be present on the node, I do not find it under `/var/log/`.

My main concern is: How do I get rid of this pod/node to a healthy one? One idea is to set the node to a state where it won't accept new pods, scale my cluster up by 1 node, and then start to delete pods. My idea is that the replication-sets connected to those pods will then recreate pods, but on the new node. When all my pods are moved to the new node, I can simply delete this faulty node. Do you think this is a good approach?

**Comment by MrHohn** [comment]
> I'm not 100% sure which kubelet.log file you mean. If it is a file that should be present on the node, I do not find it under /var/log/.

@Jacobh2 Regarding the logs, you may need to use `journalctl`. Could you try `journalctl -u kubelet`?

Ref: https://kubernetes.io/docs/tasks/debug-application-cluster/debug-cluster/

**Comment by yujuhong** [comment]
> @yujuhong I filtered out all the containers that did not match on "kube-proxy" and ended up with these:

@Jacobh2 could you check the `touch-lock` as @resouer suggested? My guess is the init container was run and *removed* by someone (e.g., kubelet). I'd like to confirm that.
We *might* be able to figure out who (and under what circumstances) removed the init container by looking at the logs. 
One possibility is that kubelet aggressively removed the exited init container to reclaim the resource. @dashpole is this possible with the current implementation?

> My main concern is: How do I get rid of this pod/node to a healthy one? One idea is to set the node to a state where it won't accept new pods, scale my cluster up by 1 node, and then start to delete pods. My idea is that the replication-sets connected to those pods will then recreate pods, but on the new node. When all my pods are moved to the new node, I can simply delete this faulty node. Do you think this is a good approach?

Most likely if you force kube-proxy to restart, the node will go back to normal. For example, you can force this by running `docker ps stop f923c7cb38b5`.

**Comment by Jacobh2** [comment]
I ran `journalctl -u kubelet > kubelet.log` at the affected node and downloaded that file. Opened it and searched for `touch-lock` and unfortunately it gave zero hits. Is there anything else I can do to find some logs that will help with this?

So just to clarify: If I ssh into the affected node and stop the docker container, kubernetes will see this and restart the correct container? Meaning, I do not stop/delete/restart anything using `kubectl`?

**Comment by dashpole** [comment]
Yes, this is possible with the current implementation.  I will post a fix for this.

**Comment by yujuhong** [comment]
> I ran journalctl -u kubelet > kubelet.log at the affected node and downloaded that file. Opened it and searched for touch-lock and unfortunately it gave zero hits. Is there anything else I can do to find some logs that will help with this?

@dashpole, is there any log message from eviction manager that could help us confirm the cause?


> So just to clarify: If I ssh into the affected node and stop the docker container, kubernetes will see this and restart the correct container? Meaning, I do not stop/delete/restart anything using kubectl?

Kubelet should attempt to restart the containers, and I believe it should start the init container again as well.

**Comment by dashpole** [comment]
We should see the following if this was aggressive garbage collection triggered by disk pressure:
`eviction manager: attempting to delete unused containers`
followed closely by:
`Removing container ` + container id

**Comment by Jacobh2** [comment]
@dashpole I searched for `Removing container` in the log and found this:
```
Sep 14 14:23:28 gke-<CLUSTER-NAME>-default-pool-e103106a-jv06 kubelet[1329]: W0914 14:23:28.356160    1329 kubenet_linux.go:488] Failed to delete container from kubenet: Error removing container from network: failed to Statfs "/proc/7823/ns/net": no such file or directory
...
Sep 15 07:12:40 gke-<CLUSTER-NAME>-default-pool-e103106a-jv06 kubelet[1329]: W0915 07:12:40.198496    1329 kubenet_linux.go:488] Failed to delete container from kubenet: Error removing container from network: failed to Statfs "/proc/32729/ns/net": no such file or directory
```

Then I searched for `eviction manager: attempting to delete unused containers` and found this:
```
Sep 17 14:03:47 gke-<CLUSTER-NAME>-default-pool-e103106a-jv06 kubelet[1329]: I0917 14:03:47.872562    1329 helpers.go:1057] eviction manager: attempting to delete unused containers
...
Sep 20 10:46:04 gke-<CLUSTER-NAME>-default-pool-e103106a-jv06 kubelet[1329]: I0920 10:46:04.732578    1329 helpers.go:1057] eviction manager: attempting to delete unused containers
...
Sep 21 08:48:28 gke-<CLUSTER-NAME>-default-pool-e103106a-jv06 kubelet[1329]: I0921 08:48:28.635371    1329 helpers.go:1057] eviction manager: attempting to delete unused containers
...
Sep 22 08:21:21 gke-<CLUSTER-NAME>-default-pool-e103106a-jv06 kubelet[1329]: I0922 08:21:21.921772    1329 helpers.go:1057] eviction manager: attempting to delete unused containers
...
Sep 22 13:34:23 gke-<CLUSTER-NAME>-default-pool-e103106a-jv06 kubelet[1329]: I0922 13:34:23.133517    1329 helpers.go:1057] eviction manager: attempting to delete unused containers
...
Sep 26 11:26:32 gke-<CLUSTER-NAME>-default-pool-e103106a-jv06 kubelet[1329]: I0926 11:26:32.513962    1329 helpers.go:1057] eviction manager: attempting to delete unused containers
```
Directly following each row in the log of this, there was also this logged:
`Sep 17 14:03:48 gke-<CLUSTER-NAME>-default-pool-e103106a-jv06 kubelet[1329]: I0917 14:03:48.067510    1329 helpers.go:1067] eviction manager: attempting to delete unused images`

And this can be it: My nodes "only" have 32GB sized disks, because I took for granted that since there will be nothing stored on these disks, they don't have to be so big. The only thing stored should be the docker-images, which I assumed was being deleted in some LRU policy or similar.

**Comment by Jacobh2** [comment]
> And this can be it: My nodes "only" have 32GB sized disks, because I took for granted that since there will be nothing stored on these disks, they don't have to be so big. The only thing stored should be the docker-images, which I assumed was being deleted in some LRU policy or similar.

I ssh into the affected node again and ran `df -h` and there was at most used 37% of disk-usage, so it doesn't look like it was disc-pressure?

**Comment by Jacobh2** [comment]
I also ran `docker stop <container-hash>` on the node, but running `kubectl -n kube-system get pods` still shows:
```
NAME                                                        READY     STATUS     RESTARTS   AGE
fluentd-gcp-v2.0-wdsg2                                      2/2       Running    0          3m
kube-proxy-gke-<CLUSTER-NAME>-default-pool-e103106a-jv06    0/1       Init:0/1   0          13d
```
But interestingly, the `fluentd-gcp-v2.0-wdsg2` seems to have been restarted recentrly!

**Comment by resouer** [comment]
@Jacobh2 As you mentioned above, the iptables on this node has been updated (means kube-proxy container has start to work)? While initContainer in same Pod is still `Waiting`, am I right?

**Comment by Jacobh2** [comment]
@resouer The Init Containers in the kube-proxy pod still has state `Waiting` with reason `PodInitializing`. But I guess the iptables has been updated since the fluentd-pod has been restarted.

**Comment by dashpole** [comment]
The `Removing container` line should come from here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_container.go#L813.
However, it looks like it is logged with V(4), so it may not appear in your logs...

**Comment by dashpole** [comment]
`eviction manager: attempting to delete unused containers` is only triggered under disk pressure, so you definitely had disk pressure at some point in the past.  But we do not know for sure if the container was removed by kubelet garbage collection.

**Comment by Jacobh2** [comment]
Ok, interesting. Does this mean that you cannot have a node-disk at 32 GB, or the fact that I now appear to have ~63% free that it apparently succeeded to clear unneeded files, but that the disk pressure also triggered this problem?

**Comment by yujuhong** [comment]
I think it's the later. The node was under disk pressure, and the GC kicked in and cleaned up the space. The only thing was that there was a bug that caused the exited init container to be cleaned up in the process too.  @dashpole's PR should fix that.

**Comment by Jacobh2** [comment]
Cool! Thank you for all your help and information! Hope to see Google cherry-picking this fix into 1.7.X soon!

**Comment by benbonnet** [comment]
Hello, sorry to bother for this one, experiencing the same issue. One of my kube-proxy pod is stuck pending; same state as shown in the original post

The disks weren't pressured, but a deployment was not limited and went out of its mind (the cpu bursted to 120% or more for some hours), I expect it to be the reason (with no way of being sure of it). Anyhow, some trouble happened,  and here i am with some un-restartable kind-of-critical pod; reading again here, i still hardly understand where to kick it in order to have it up; could you help/remind about the command to run ?
